internal/xds/rbac/rbac_engine.go

```diff
@@ -15,68 +15,210 @@
 		if err != nil {
 			return nil, err
 		}
 		policies[name] = matcher
 	}
-	return &Engine{policies: policies}, nil
+	return &engine{
+		policies: policies,
+		action:   action,
+	}, nil
 }
 
-// RPCData wraps data pulled from an incoming RPC that the RBAC engine needs to
-// find a matching policy.
-type RPCData struct {
-	// MD is the HTTP Headers that are present in the incoming RPC.
-	MD metadata.MD
-	// PeerInfo is information about the downstream peer.
-	PeerInfo *peer.Peer
-	// FullMethod is the method name being called on the upstream service.
-	FullMethod string
-	// DestinationPort is the port that the RPC is being sent to on the
-	// server.
-	DestinationPort uint32
-	// DestinationAddr is the address that the RPC is being sent to.
-	DestinationAddr net.Addr
-	// PrincipalName is the name of the downstream principal. If set, the URI
-	// SAN or DNS SAN in that order is used from the certificate, otherwise the
-	// subject field is used. If unset, it applies to any user that is
-	// authenticated.
-	PrincipalName string
-}
+var errPolicyNotFound = errors.New("a matching policy was not found")
 
-// FindMatchingPolicy determines if an incoming RPC matches a policy. On a
-// successful match, it returns the name of the matching policy and a true
-// boolean to specify that there was a matching policy found.
-func (r *Engine) FindMatchingPolicy(data *RPCData) (string, bool) {
+// findMatchingPolicy determines if an incoming RPC matches a policy. On a
+// successful match, it returns the name of the matching policy and a nil error
+// to specify that there was a matching policy found.  It returns an error in
+// the case of not finding a matching policy.
+func (r *engine) findMatchingPolicy(rpcData *rpcData) (string, error) {
 	for policy, matcher := range r.policies {
-		if matcher.match(data) {
-			return policy, true
+		if matcher.match(rpcData) {
+			return policy, nil
 		}
 	}
-	return "", false
+	return "", errPolicyNotFound
+}
+
+// newRPCData takes a incoming context (should be a context representing state
+// needed for server RPC Call with metadata, peer info (used for source ip/port
+// and TLS information) and connection (used for destination ip/port) piped into
+// it) and the method name of the Service being called server side and populates
+// an rpcData struct ready to be passed to the RBAC Engine to find a matching
+// policy.
+func newRPCData(ctx context.Context, methodName string) (*rpcData, error) {
+	// The caller should populate all of these fields (i.e. for empty headers,
+	// pipe an empty md into context).
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, errors.New("missing metadata in incoming context")
+	}
+
+	pi, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, errors.New("missing peer info in incoming context")
+	}
+
+	// The connection is needed in order to find the destination address and
+	// port of the incoming RPC Call.
+	conn, ok := getConnection(ctx)
```

Comment:
It's totally fine to not have a second return value here. Just check for `conn == nil` and return an error in that case.

Reply:
Sure thing. Got rid of bool return value.